### PR TITLE
CM-546: Type Checking And Casting in C#

### PR DESCRIPTION
**Pull request by dgarcia202**

### CODE REVIEW

File: csharp-basic-topics/TypeCheckingAndCastingInCSharp/TypeCheckingAndCastingInCSharp.Tests/TypeConversionTests.cs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            var floatAsBytes = BitConverter.GetBytes(fraction);
+
+            var format = "{0,12}{1,15}";
+            Console.WriteLine(format, "Base Type", "Bytes");
+            Console.WriteLine(format, "=========", "=====");
+            Console.WriteLine(format, integer, BitConverter.ToString(integerAsBytes));
+            Console.WriteLine(format, fraction, BitConverter.ToString(floatAsBytes));
+
+            // Assert
+            Assert.Equal("   Base Type          Bytes" + Environment.NewLine + "   =========          =====" + Environment.NewLine + "        1024    00-04-00-00" + Environment.NewLine + "        12.5    00-00-48-41" + Environment.NewLine, sw.ToString());
```
Comment (CodeMazeBlog) [inline csharp-basic-topics/TypeCheckingAndCastingInCSharp/TypeCheckingAndCastingInCSharp.Tests/TypeConversionTests.cs:143]: Please no SetOut method for tests. Just had a discussion with Marinko the other day. It's unreliable and hard to create.
This is easy to test without it:
image

File: csharp-basic-topics/TypeCheckingAndCastingInCSharp/TypeCheckingAndCastingInCSharp.Tests/TypeTestingTests.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            Assert.Equal(typeof(string), stringValue.GetType());
+            Assert.Equal(typeof(float), singleValue.GetType());
+        }
+
+        [Fact]
+        public void WhenTypeTestingAtRunTime_ThenUseGeTypeAndTypeof()
+        {
+            // arrange 
+            var sw = new StringWriter();
+            Console.SetOut(sw);
```
Comment (CodeMazeBlog) [inline csharp-basic-topics/TypeCheckingAndCastingInCSharp/TypeCheckingAndCastingInCSharp.Tests/TypeTestingTests.cs:51]: Please create an alternative without SetOut.

File: csharp-basic-topics/TypeCheckingAndCastingInCSharp/TypeCheckingAndCastingInCSharp.Tests/TypeConversionTests.cs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        public void WhenOverloadingCastingOperator_ThenImplicitAndExplicitConversionsWork()
+        {
+            var currency = new Currency("$", 1950.75M);
+            decimal decimalValue = currency;    // decimalValue contains 1950.75
+
+            var anotherDecimal = 775.2M;
+            var convertedCurrency = (Currency)anotherDecimal;  // convertedCurrency prints "775.20 $"
+
+            Assert.Equal(1950.75M, decimalValue);
+            Assert.Equal("775.20 $", convertedCurrency.ToString());
```
Comment (CodeMazeBlog) [inline csharp-basic-topics/TypeCheckingAndCastingInCSharp/TypeCheckingAndCastingInCSharp.Tests/TypeConversionTests.cs:156]: Change this as per review as well.

File: csharp-basic-topics/TypeCheckingAndCastingInCSharp/TypeCheckingAndCastingInCSharp.Tests/TypeConversionTests.cs (hunk 1 above)
Comment (dgarcia202) [reply csharp-basic-topics/TypeCheckingAndCastingInCSharp/TypeCheckingAndCastingInCSharp.Tests/TypeConversionTests.cs]: fixed, following your example

File: csharp-basic-topics/TypeCheckingAndCastingInCSharp/TypeCheckingAndCastingInCSharp.Tests/TypeTestingTests.cs (hunk 2 above)
Comment (dgarcia202) [reply csharp-basic-topics/TypeCheckingAndCastingInCSharp/TypeCheckingAndCastingInCSharp.Tests/TypeTestingTests.cs]: done

File: csharp-basic-topics/TypeCheckingAndCastingInCSharp/TypeCheckingAndCastingInCSharp.Tests/TypeConversionTests.cs (hunk 3 above)
Comment (dgarcia202) [reply csharp-basic-topics/TypeCheckingAndCastingInCSharp/TypeCheckingAndCastingInCSharp.Tests/TypeConversionTests.cs]: done

File: csharp-basic-topics/TypeCheckingAndCastingInCSharp/TypeCheckingAndCastingInCSharp.Tests/TypeConversionTests.cs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+            // act
+            var integerAsBytes = BitConverter.GetBytes(integer);
+            var floatAsBytes = BitConverter.GetBytes(fraction);
+
+            var format = "{0,12}{1,15}";
+            Console.WriteLine(format, "Base Type", "Bytes");
+            Console.WriteLine(format, "=========", "=====");
+            Console.WriteLine(format, integer, BitConverter.ToString(integerAsBytes));
+            Console.WriteLine(format, fraction, BitConverter.ToString(floatAsBytes));
```
Comment (CodeMazeBlog) [inline csharp-basic-topics/TypeCheckingAndCastingInCSharp/TypeCheckingAndCastingInCSharp.Tests/TypeConversionTests.cs:138]: We don't need these now.

Comment (dgarcia202) [reply csharp-basic-topics/TypeCheckingAndCastingInCSharp/TypeCheckingAndCastingInCSharp.Tests/TypeConversionTests.cs]: ok, removed

**Pull request merged**